homeassistant/components/binary_sensor/w800rf32.py

```diff
@@ -0,0 +1,179 @@
+from homeassistant.components.w800rf32 import (
+    DOMAIN, CONF_FIRE_EVENT, CONF_OFF_DELAY)
+from homeassistant.const import (
+    ATTR_ENTITY_ID, ATTR_STATE,
+    CONF_DEVICE_CLASS, CONF_NAME,
+    ATTR_NAME, CONF_DEVICES)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import event as evt
+from homeassistant.util import dt as dt_util
+from homeassistant.helpers.dispatcher import (dispatcher_connect)
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['w800rf32']
+DEVICE_TYPE = 'W800rf32BinarySensor'
+
+ATTR_FIRE_EVENT = 'fire_event'
+EVENT_BUTTON_PRESSED = 'button_pressed'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_DEVICES, default={}): {
+        cv.string: vol.Schema({
+            vol.Optional(CONF_NAME): cv.string,
+            vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
+            vol.Optional(CONF_FIRE_EVENT, default=False): cv.boolean,
+            vol.Optional(CONF_OFF_DELAY):
+            vol.Any(cv.time_period, cv.positive_timedelta)
+        })
+    },
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Binary Sensor platform to w800rf32."""
+    import W800rf32 as w800rf32mod
+    binary_sensors = []
+    # device_id --> "c1 or a3" X10 device. entity (type dictionary)
+    # --> name, device_class etc
+    for device_id, entity in config[CONF_DEVICES].items():
+
+        if device_id in hass.data[DOMAIN]['entities']:
+            continue
+
+        _LOGGER.debug("Add %s w800rf32.binary_sensor (class %s)",
+                      entity[ATTR_NAME], entity.get(CONF_DEVICE_CLASS))
+
+        device = W800rf32BinarySensor(
+            device_id, entity.get(CONF_NAME), entity.get(CONF_DEVICE_CLASS),
+            entity[CONF_FIRE_EVENT], entity.get(CONF_OFF_DELAY))
+
+        hass.data[DOMAIN]['entities'][device_id] = device
+        binary_sensors.append(device)
+
+    add_entities(binary_sensors)
+
+    def binary_sensor_update(event):
+        """Call for control updates from the w800rf32 gateway."""
+        sensor = None
+
+        if not isinstance(event, w800rf32mod.W800rf32Event):
+            return
+
+        dev_id = event.device.lower()
+        command = event.command
+
+        #
+        # # get the name, ex: motion_hall
+        if dev_id in hass.data[DOMAIN]['entities']:
+            sensor = hass.data[DOMAIN]['entities'][dev_id]
+
+        if sensor is None:
+            return
+
+        if not isinstance(sensor, W800rf32BinarySensor):
+            return
+
+        _LOGGER.debug(
+            "BinarySensor update (Device ID: %s Command %s ...)",
+            dev_id, command)
+
+        # Update the w800rf32 device state
+        if command in ('On', 'Off'):
+            is_on = command == 'On'
+            sensor.update_state(is_on)
+
+        # Fire event
+        if sensor.should_fire_event:
+            sensor.hass.bus.fire(
```

Comment:
Entities in platforms shouldn't normally fire events. Entities already fire state change events.

Its ok to move this to the component and fire the event after signaling the device id. We can't configure entities to fire events in that case though.

@balloob what do you think?

Reply:
Correct, sensors should never fire events. If a binary sensor needs to fire an event to be useable, it shouldn't be a binary sensor to begin with and just be an event.